### ISSUE TITLE
Invalid file names are generated because of test names

### ISSUE THREAD

**Issue by TWiStErRob** [mannodermaus/android-junit5#263]
I'm having these tests in `src/androidTest/kotlin`:
```kotlin
class ColorUtilsKtTest {

	class ReplaceAlphaFrom(
		val useCase: String,
		@ColorInt val receiver: Int,
		@ColorInt val param: Int,
		@ColorInt val expected: Int,
	)

	@TestFactory fun replaceAlphaFrom() =
		listOf(
			ReplaceAlphaFrom(
				"Replaces alpha keeping RGB",
				Color.argb(0x12, 0x34, 0x56, 0x78),
				Color.argb(0x87, 0x65, 0x43, 0x21),
				Color.argb(0x87, 0x34, 0x56, 0x78),
			),
			ReplaceAlphaFrom(
				"Adds only alpha",
				Color.argb(0x00, 0x00, 0x00, 0x00),
				Color.argb(0xAB, 0x00, 0x00, 0x00),
				Color.argb(0xAB, 0x00, 0x00, 0x00),
			),
			ReplaceAlphaFrom(
				"Clears alpha",
				Color.argb(0x12, 0x34, 0x56, 0x78),
				Color.argb(0x00, 0x00, 0x00, 0x00),
				Color.argb(0x00, 0x34, 0x56, 0x78),
			),
		).map {
			dynamicTest(it.useCase) {
				val result = it.receiver.replaceAlphaFrom(it.param)

				assertThat(it.expected, equalTo(result))
			}
		}
}
```
When running them in Android Studio I get
![image](https://user-images.githubusercontent.com/2906988/157620653-f957777a-b72a-4bf5-9677-7d87f0d96341.png)
which is beautiful, well done!

I have the same structure in `src/test/kotlin` too, that looks like this:
![image](https://user-images.githubusercontent.com/2906988/157620847-f5f4826b-0ce4-4f3b-bba9-51692bc6d7cc.png)

At this point I can see that there's a limitation of nesting test containers in androidTest and you worked around it by concat-ing. I've identified this line to be responsible:

https://github.com/mannodermaus/android-junit5/blob/9136759f3685b4def2d4124948aa62eb5d5c233c/instrumentation/runner/src/main/kotlin/de/mannodermaus/junit5/internal/runners/AndroidJUnitPlatformTestTree.kt#L53

---

Now, onto the issue...
When running on AGP 7.1 (`gradlew :feature:base:connectedCheck`) there are some new features to dump the logcat and other files, see what is the output here:
![image](https://user-images.githubusercontent.com/2906988/157621765-69c942f6-0a37-4230-a827-9c9a8dfe21ae.png)

This is on Windows, where there are [reserved characters](https://docs.microsoft.com/en-us/windows/win32/fileio/naming-a-file#naming-conventions) in file names (see also [wiki](https://en.wikipedia.org/wiki/Filename#Reserved_characters_and_words). You can see this in action, because `build/outputs/androidTest-results/connected/API_29(AVD) - 10/logcat-net.twisterrob.colorfilters.android.ColorUtilsKtTest-replaceAlphaFrom` file name is truncated at the `:`, and actually there's only 1 file, not 3. ([AGP issue](https://issuetracker.google.com/issues/223643506))

When running the same on Mac (CI) I get the full name: `build/outputs/androidTest-results/connected/test(AVD) - 10/logcat-net.twisterrob.colorfilters.android.ColorUtilsKtTest-replaceAlphaFrom: Adds only alpha.txt`, but when trying to upload artifacts I get an error because of the "invalid" character: https://github.com/TWiStErRob/net.twisterrob.colorfilters/runs/5487958432?check_suite_focus=true#step:7:17
([AGP issue](https://issuetracker.google.com/issues/223643506))

![image](https://user-images.githubusercontent.com/2906988/157626185-9bd8aea1-599d-467b-ae40-52ed3b0322c7.png)

At this point I would like to ask to make it configurable what the joiner characters are, e.g. I would configure it to ` - ` to prevent this error. Related android-test issue: https://github.com/android/android-test/issues/1314

Repro: https://github.com/TWiStErRob/repros/tree/master/agp/junit5-logcat-filenames

**Comment by mannodermaus** [comment]
Thanks for bringing this up and the thorough description! 🙏 Ideally we'd like to retain the colon characters for test reporting, but sanitize the handoff to this Logcat task for generation of its file names. Gotta dive into how this new implementation of `connectedCheck` pulls these files.

**Comment by mannodermaus** [comment]
I have entered the rabbit hole that is the Unified Test Platform for Android (UTP) and wanted to share some initial findings. UTP was apparently introduced with AGP 7.0 and takes care of the test reporting for instrumentation tests through a variety of plugins ([here's a list of all the Maven artifacts for these plugins](https://maven.google.com/web/index.html?q=utp)).

I found `UtpTestUtils.kt` inside AGP 7.1, where the output folder for the executing device is created (the one that stores `test-result.pb`). The remaining contents of the folder are handled by various plugins, including one for Logcat. Inside this Logcat dependency, the main plugin class is generating the problematic file names directly from the contents of Logcat itself ([source link 1](https://android.googlesource.com/platform/tools/base/+/d5534e88487c465d9b9643fc2893d87523f85247/utp/android-test-plugin-host-logcat/src/main/java/com/android/tools/utp/plugins/host/logcat/AndroidTestLogcatPlugin.kt#172), [source link 2](https://android.googlesource.com/platform/tools/base/+/d5534e88487c465d9b9643fc2893d87523f85247/utp/android-test-plugin-host-logcat/src/main/java/com/android/tools/utp/plugins/host/logcat/AndroidTestLogcatPlugin.kt#101)):

```kotlin
private fun parseLine(line: String) {
    val testPackageClassAndMethodNames = line.split("TestRunner: started: ")[1].split("(")
    val testMethod = testPackageClassAndMethodNames[0].trim()
    val testPackageAndClass = testPackageClassAndMethodNames[1].removeSuffix(")")
    val tempFileName = generateLogcatFileName(testPackageAndClass, testMethod) // <----- !!
    tempLogcatFile = File(tempFileName)
    logcatFilePaths.add(tempFileName)
    tempLogcatWriter = tempLogcatFile.outputStream().bufferedWriter()
}

private fun generateLogcatFileName(
        testPackageAndClass: String,
        testMethod: String
) = File(outputDir, "logcat-$testPackageAndClass-$testMethod.txt").absolutePath // <----- !!
```

I can't really change the output of Logcat, though - else the IDE reporting and other places would start to look messy again. I wonder if the fix could be applied from AGP's side, cleaning the generated file names of any illegal symbols. We managed to fix other upstream bugs in the past so maybe this could be another attempt at one. 🤔

**Comment by TWiStErRob** [comment]
I opened a bug for AGP too, would you mind commenting on the issue with your findings?

In the meantime please consider making a change to allow for at least making it work in some way. A configuration param that's off by default, or even just mutable private field that can be set with reflection 🥺. "Ugly working" is better than "pretty, but broken".

**Comment by mannodermaus** [comment]
I copied my findings to the ticket on the Google issue tracker, thanks for the reminder. As for a workaround from android-junit5's side, the easiest thing I can think of is to not use colons in the test output at all. Replacing the `: ` separator with ` - ` will change the reporting in the IDE, but should stop breaking the file generation of the Logcat host plugin.

|Before|After|
|---|---|
|<img width="700" alt="Screen Shot 2022-04-10 at 19 45 13" src="https://user-images.githubusercontent.com/3897705/162632531-ef32cc3d-083b-4eee-8f6a-03968d052279.png">|<img width="700" alt="Screen Shot 2022-04-10 at 19 47 31" src="https://user-images.githubusercontent.com/3897705/162632534-b68ac3e1-ad3a-4c74-8eff-399f9f87d029.png">|

The remaining punctuation in these file names (comma, square brackets) should not pose a problem for Windows, right?

**Comment by TWiStErRob** [comment]
Confirmed that Windows has no issues with these characters:
`-. ()[]`

Screenshots of `logcat-de.mann.sample.Test.f(RepetitionInfo, AcitivtyScenario) - [1] repetition 1 of 3-repeatedTestExample` in Windows Explorer (Win 10):
![image](https://user-images.githubusercontent.com/2906988/162633370-f2c1d5fa-eda4-48eb-b1bd-ce8e0dfcae1d.png)

**Comment by TWiStErRob** [comment]
Note: there's something weird in your screenshots, the open `(` is missing.

**Comment by TWiStErRob** [comment]
Btw, I tried to type a `:` and Windows told me this:
![image](https://user-images.githubusercontent.com/2906988/162633634-3360900e-d513-453e-89e1-1d88bebc79bf.png)

Mind you, this means that `&`, `;`, `{`/`}`, `%`, `=`, `'`, `!` are all valid characters.

![image](https://user-images.githubusercontent.com/2906988/162633707-1c21edf0-3ea5-4db2-9ac9-c48cc5d46dd2.png)

**Comment by mannodermaus** [comment]
Thanks for confirming! 

> Note: there's something weird in your screenshots, the open `(` is missing.

Yeah, I noticed that too. It's likely a side effect of this Logcat plugin splitting the test name by `(` and just using the first two elements of the result, which doesn't bode well when the test name itself already has parentheses.

In any case, version `1.3.1` of the instrumentation test libraries should have this fix. If you need it now, please use their `-SNAPSHOT` instead!

**Comment by TWiStErRob** [comment]
The effect of this paren split is that two tests named the same in different classes would be in the same logcat files (test class name missing). This is less likely though.

I had a quick scan of https://developer.android.com/jetpack/androidx/releases/test#version_131_2 but can't see the fix listed.

**Comment by mannodermaus** [comment]
Ah no, sorry - I was referring to the fix on android-junit5's side:

```kotlin
dependencies {
  androidTestImplementation("de.mannodermaus.junit5:android-test-core:1.3.1-SNAPSHOT")
  androidTestRuntimeOnly("de.mannodermaus.junit5:android-test-runner:1.3.1-SNAPSHOT")
}
```

**Comment by mannodermaus** [comment]
Released in `1.4.0` of the instrumentation libraries
